lib/Sema/CSSolver.cpp

```diff
@@ -1115,6 +1115,11 @@ ConstraintSystem::solveImpl(Expr *&expr,
   // constraint.
   if (convertType) {
     auto constraintKind = ConstraintKind::Conversion;
+    
+    if (getContextualTypePurpose() == CTP_ReturnStmt
+        && Options.contains(ConstraintSystemFlags::UnderlyingTypeForOpaqueReturnType))
```

Comment:
I wonder - do we really need a new flag or we could detect this situation based on `convertType` itself, because it always would be `OpaqueTypeArchetypeType` right?

Reply:
I think we normally want to treat the opaque type and underlying type as distinct outside of return statements.

Reply:
Sorry, what I meant was:

```c++
if (getContextualTypePurpose() == CTP_ReturnStmt &&
    covertType->isa<OpaqueTypeArchetypeType>())
  constraintKind = ConstraintKind::OpaqueUnderlyingType;
```

Reply:
The docs for ContextualTypePurpose suggested that it was only intended as a discriminator for diagnostics, so I wasn't sure whether it was OK to key other constraint system behavior solely on it. I can do what you suggested if you think it's better.

Reply:
I think it was intended like that originally but it would be totally okay to use it for this case as well to avoid adding new flags.